File: backend_go/internal/handlers/budgets.go
package handlers

import (
	"household-go/internal/database"
	"household-go/internal/models"
	"household-go/internal/utils"
	"log"

	"github.com/gin-gonic/gin"
)

// BudgetsHandler handles budget queries and overrides.
type BudgetsHandler struct{}

func NewBudgetsHandler() *BudgetsHandler { return &BudgetsHandler{} }

// GetMonthBudget returns all expense categories with spending vs budget for a calendar month.
// GET /api/app/budget/month?householdId=X&year=Y&month=M
func (h *BudgetsHandler) GetMonthBudget(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)

	householdID := c.Query("householdId")
	year := c.Query("year")
	month := c.Query("month")

	if householdID == "" || year == "" || month == "" {
		utils.BadRequest(c, "householdId, year, and month are required")
		return
	}

	if err := verifyMembership(householdID, appUser.ID); err != nil {
		utils.Forbidden(c, "Not a member of this household")
		return
	}

	yearInt := parseInt(year)
	monthInt := parseInt(month)

	// Calendar month boundaries: first day to last day of the month.
	// time.Date handles month overflow — time.Date(2024, 13, 1,...) = 2025-01-01.
	// monthStart = first second of the month
	// monthEnd   = last second of the month (day 0 of next month = last day of this month)

	// We use raw SQL for the SUM aggregation — more efficient than N+1 Go loops.
	// GORM's Raw() lets you write raw SQL while still getting typed scan results.
	type budgetRow struct {
		CategoryID uint    `gorm:"column:id"`
		Spent      float64 `gorm:"column:spent"`
	}

	// First, load all categories for this household
	var categories []models.ExpenseCategory
	if err := database.DB.Where("household_id = ?", householdID).
		Order("sort_order ASC").Find(&categories).Error; err != nil {
		utils.InternalError(c, "Failed to retrieve month budget")
		return
	}

	// Load budget overrides for this month
	var overrides []models.CategoryBudgetOverride
	database.DB.Where("household_id = ? AND year = ? AND month = ?", householdID, yearInt, monthInt).
		Find(&overrides)

	// Build override map: categoryID → override record
	// Maps in Go are reference types initialized with make() or a literal.
	overrideMap := make(map[uint]models.CategoryBudgetOverride)
	for _, o := range overrides {
		overrideMap[o.ExpenseCategoryID] = o
	}

	// Aggregate spending per category in a single query instead of N queries.
	// This is the key optimization vs the original Node code which does one query per category.
	// GORM's Raw() returns a *gorm.DB; Scan() executes it and fills the slice.
	var spendRows []budgetRow
	database.DB.Raw(`
		SELECT expense_category_id AS id, COALESCE(SUM(amount), 0) AS spent
		FROM expenses
		WHERE household_id = ?
		  AND date_time >= DATE(CONCAT(?, '-', ?, '-01'))
		  AND date_time < DATE(CONCAT(?, '-', ?+1, '-01'))
		GROUP BY expense_category_id
	`, householdID, year, month, year, month).Scan(&spendRows)

	spendMap := make(map[uint]float64)
	for _, row := range spendRows {
		spendMap[row.CategoryID] = row.Spent
	}

	// Build the response slice
	type categoryResult struct {
		ID              uint     `json:"id"`
		Name            string   `json:"name"`
		NameHe          *string  `json:"nameHe"`
		Icon            *string  `json:"icon"`
		Color           *string  `json:"color"`
		BaseBudget      *float64 `json:"baseBudget"`
		Override        *float64 `json:"override"`
		EffectiveBudget *float64 `json:"effectiveBudget"`
		Spent           float64  `json:"spent"`
		Result          *float64 `json:"result"`
	}

	result := make([]categoryResult, 0, len(categories))
	for _, cat := range categories {
		spent := spendMap[cat.ID]
		override, hasOverride := overrideMap[cat.ID]

		var overrideAmount *float64
		if hasOverride {
			a := override.Amount
			overrideAmount = &a
		}

		var effectiveBudget *float64
		if overrideAmount != nil {
			effectiveBudget = overrideAmount
		} else {
			effectiveBudget = cat.MonthlyBudget
		}

		var budgetResult *float64
		if effectiveBudget != nil {
			r := spent - *effectiveBudget
			budgetResult = &r
		}

		result = append(result, categoryResult{
			ID:              cat.ID,
			Name:            cat.Name,
			NameHe:          cat.NameHe,
			Icon:            cat.Icon,
			Color:           cat.Color,
			BaseBudget:      cat.MonthlyBudget,
			Override:        overrideAmount,
			EffectiveBudget: effectiveBudget,
			Spent:           spent,
			Result:          budgetResult,
		})
	}

	utils.OK(c, gin.H{"success": true, "categories": result})
}

// SetBaseBudget sets the default monthly budget on an expense category.
// PUT /api/app/budget/base
func (h *BudgetsHandler) SetBaseBudget(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)

	var body struct {
		ExpenseCategoryID uint     `json:"expenseCategoryId" binding:"required"`
		HouseholdID       uint     `json:"householdId"       binding:"required"`
		Amount            *float64 `json:"amount"` // pointer: allows sending 0 or null
	}
	if err := c.ShouldBindJSON(&body); err != nil || body.Amount == nil {
		utils.BadRequest(c, "expenseCategoryId, householdId, and amount are required")
		return
	}

	if err := verifyMembership(formatUint(body.HouseholdID), appUser.ID); err != nil {
		utils.Forbidden(c, "Not a member of this household")
		return
	}

	var category models.ExpenseCategory
	if err := database.DB.Where("id = ? AND household_id = ?", body.ExpenseCategoryID, body.HouseholdID).
		First(&category).Error; err != nil {
		utils.NotFound(c, "Category not found")
		return
	}

	var newBudget *float64
	if *body.Amount >= 0 {
		newBudget = body.Amount
	} // nil if negative — clears the budget

	database.DB.Model(&category).Update("monthly_budget", newBudget)
	utils.OK(c, gin.H{"success": true})
}

// SetOverride upserts a CategoryBudgetOverride for a specific year/month.
// PUT /api/app/budget/override
func (h *BudgetsHandler) SetOverride(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)

	var body struct {
		ExpenseCategoryID uint    `json:"expenseCategoryId" binding:"required"`
		HouseholdID       uint    `json:"householdId"       binding:"required"`
		Year              int     `json:"year"              binding:"required"`
		Month             int     `json:"month"             binding:"required"`
		Amount            float64 `json:"amount"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		utils.BadRequest(c, "expenseCategoryId, householdId, year, month, and amount are required")
		return
	}

	if err := verifyMembership(formatUint(body.HouseholdID), appUser.ID); err != nil {
		utils.Forbidden(c, "Not a member of this household")
		return
	}

	// GORM's Save() does an upsert when the primary key is 0 (new record).
	// But for a composite-key upsert we use FirstOrCreate + Update.
	// Equivalent to Sequelize's upsert({ expenseCategoryId, householdId, year, month, amount }).
	var override models.CategoryBudgetOverride
	result := database.DB.Where(models.CategoryBudgetOverride{
		ExpenseCategoryID: body.ExpenseCategoryID,
		HouseholdID:       body.HouseholdID,
		Year:              body.Year,
		Month:             body.Month,
	}).FirstOrCreate(&override)

	if result.Error != nil {
		log.Printf("SetOverride error: %v", result.Error)
		utils.InternalError(c, "Failed to set budget override")
		return
	}

	database.DB.Model(&override).Update("amount", body.Amount)
	override.Amount = body.Amount

	utils.OK(c, gin.H{"success": true, "override": override})
}

// GetSpendingByWeek groups expenses by ISO week within a calendar month.
// GET /api/app/budget/by-week?householdId=X&year=Y&month=M
func (h *BudgetsHandler) GetSpendingByWeek(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)

	householdID := c.Query("householdId")
	year := c.Query("year")
	month := c.Query("month")
	categoryID := c.Query("expenseCategoryId")

	if householdID == "" || year == "" || month == "" {
		utils.BadRequest(c, "householdId, year, and month are required")
		return
	}

	if err := verifyMembership(householdID, appUser.ID); err != nil {
		utils.Forbidden(c, "Not a member of this household")
		return
	}

	type weekRow struct {
		ISOWeek int     `gorm:"column:iso_week"`
		Total   float64 `gorm:"column:total"`
	}

	query := `
		SELECT WEEK(date_time, 3) AS iso_week, SUM(amount) AS total
		FROM expenses
		WHERE household_id = ?
		  AND date_time >= DATE(CONCAT(?, '-', ?, '-01'))
		  AND date_time < DATE(CONCAT(?, '-', ?+1, '-01'))`

	args := []interface{}{householdID, year, month, year, month}

	if categoryID != "" {
		query += " AND expense_category_id = ?"
		args = append(args, categoryID)
	}

	query += " GROUP BY WEEK(date_time, 3) ORDER BY iso_week ASC"

	var rows []weekRow
	if err := database.DB.Raw(query, args...).Scan(&rows).Error; err != nil {
		utils.InternalError(c, "Failed to retrieve spending by week")
		return
	}

	type weekResult struct {
		WeekLabel string  `json:"weekLabel"`
		Total     float64 `json:"total"`
	}
	weeks := make([]weekResult, len(rows))
	for i, row := range rows {
		weeks[i] = weekResult{
			WeekLabel: "W" + formatInt(i+1),
			Total:     row.Total,
		}
	}

	utils.OK(c, gin.H{"success": true, "weeks": weeks})
}

// GetSpendingByMonth groups expenses by calendar month within a year range.
// GET /api/app/budget/by-month?householdId=X&year=Y&startMonth=S&endMonth=E
func (h *BudgetsHandler) GetSpendingByMonth(c *gin.Context) {
	appUser := c.MustGet("appUser").(*models.AppUser)

	householdID := c.Query("householdId")
	year := c.Query("year")
	startMonth := c.Query("startMonth")
	endMonth := c.Query("endMonth")
	categoryID := c.Query("expenseCategoryId")

	if householdID == "" || year == "" || startMonth == "" || endMonth == "" {
		utils.BadRequest(c, "householdId, year, startMonth, and endMonth are required")
		return
	}

	if err := verifyMembership(householdID, appUser.ID); err != nil {
		utils.Forbidden(c, "Not a member of this household")
		return
	}

	type monthRow struct {
		Yr    int     `gorm:"column:yr"`
		Mo    int     `gorm:"column:mo"`
		Total float64 `gorm:"column:total"`
	}

	query := `
		SELECT YEAR(date_time) AS yr, MONTH(date_time) AS mo, SUM(amount) AS total
		FROM expenses
		WHERE household_id = ?
		  AND date_time >= DATE(CONCAT(?, '-', ?, '-01'))
		  AND date_time < DATE(CONCAT(?, '-', ?+1, '-01'))`

	args := []interface{}{householdID, year, startMonth, year, endMonth}

	if categoryID != "" {
		query += " AND expense_category_id = ?"
		args = append(args, categoryID)
	}

	query += " GROUP BY YEAR(date_time), MONTH(date_time) ORDER BY yr ASC, mo ASC"

	var rows []monthRow
	if err := database.DB.Raw(query, args...).Scan(&rows).Error; err != nil {
		utils.InternalError(c, "Failed to retrieve spending by month")
		return
	}

	monthNames := [12]string{"Jan", "Feb", "Mar", "Apr", "May", "Jun",
		"Jul", "Aug", "Sep", "Oct", "Nov", "Dec"}

	type monthResult struct {
		Label string  `json:"label"`
		Total float64 `json:"total"`
	}
	months := make([]monthResult, len(rows))
	for i, row := range rows {
		months[i] = monthResult{
			Label: monthNames[row.Mo-1] + " " + formatInt(row.Yr),
			Total: row.Total,
		}
	}

	utils.OK(c, gin.H{"success": true, "months": months})
}

// verifyMembership checks that appUserID is a member of the given household.
// Returns an error if not a member (or on DB error).
func verifyMembership(householdID string, appUserID uint) error {
	var m models.HouseholdMember
	return database.DB.Where("household_id = ? AND app_user_id = ?", householdID, appUserID).
		First(&m).Error
}

// formatUint converts uint to string without importing strconv in this file.
func formatUint(n uint) string {
	return formatInt(int(n))
}

// formatInt converts int to string (simple, no import needed for this use case).
func formatInt(n int) string {
	if n == 0 {
		return "0"
	}
	buf := [20]byte{}
	pos := len(buf)
	neg := n < 0
	if neg {
		n = -n
	}
	for n > 0 {
		pos--
		buf[pos] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		pos--
		buf[pos] = '-'
	}
	return string(buf[pos:])
}

File: controllers/budgets.js
const { ExpenseCategory, CategoryBudgetOverride, Expense, RecurringExpense, Household, HouseholdMember, BudgetPlanItem, BudgetMonthConfig, sequelize } = require('../models');
const { Op, fn, col, literal } = require('sequelize');
const { getFinancialPeriodForAnchor } = require('../utils/financialCalendar');

async function getHouseholdStartDay(householdId) {
  const h = await Household.findByPk(householdId, { attributes: ['id', 'financialMonthStartDay'] });
  return h?.financialMonthStartDay ?? 10;
}

class BudgetsController {
  /**
   * App: GET /app/budget/month?householdId=X&year=Y&month=M
   * Returns all expense categories with budget and spending data for the calendar month.
   */
  async getMonthBudget(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { householdId, year, month } = ctx.query;

      if (!householdId || !year || !month) {
        ctx.status = 400;
        ctx.body = { error: 'householdId, year, and month are required' };
        return;
      }

      const membership = await HouseholdMember.findOne({
        where: { householdId, appUserId: appUser.id }
      });
      if (!membership) {
        ctx.status = 403;
        ctx.body = { error: 'Not a member of this household' };
        return;
      }

      const yearInt = parseInt(year, 10);
      const monthInt = parseInt(month, 10);

      const startDay = await getHouseholdStartDay(householdId);
      const period = getFinancialPeriodForAnchor(yearInt, monthInt, startDay);
      const monthStart = period.start;
      const monthEnd = period.end;

      const categories = await ExpenseCategory.findAll({
        where: { householdId },
        order: [['sortOrder', 'ASC']]
      });

      const overrides = await CategoryBudgetOverride.findAll({
        where: { householdId, year: yearInt, month: monthInt }
      });
      const overrideMap = {};
      for (const o of overrides) {
        overrideMap[o.expenseCategoryId] = o;
      }

      // Pre-fetch active recurring expenses for this household that apply to this month
      const allRecurring = await RecurringExpense.findAll({
        where: { householdId, isActive: true }
      });

      // Build a map of categoryId -> recurring total for this calendar month
      const recurringTotalByCategory = {};
      for (const rec of allRecurring) {
        // Check startYear/startMonth <= yearInt/monthInt
        if (rec.startYear > yearInt || (rec.startYear === yearInt && rec.startMonth > monthInt)) continue;
        const catId = rec.expenseCategoryId;
        recurringTotalByCategory[catId] = (recurringTotalByCategory[catId] || 0) + parseFloat(rec.amount);
      }

      const result = await Promise.all(categories.map(async (category) => {
        const spendRow = await Expense.findOne({
          attributes: [[fn('SUM', col('amount')), 'total']],
          where: {
            expenseCategoryId: category.id,
            dateTime: { [Op.between]: [monthStart, monthEnd] }
          },
          raw: true
        });

        const regularSpent  = parseFloat(spendRow.total) || 0;
        const recurringSpent = recurringTotalByCategory[category.id] || 0;
        const spent = regularSpent + recurringSpent;
        const override = overrideMap[category.id] || null;
        const baseBudget = category.monthlyBudget != null ? parseFloat(category.monthlyBudget) : null;
        const overrideAmount = override ? parseFloat(override.amount) : null;
        const effectiveBudget = overrideAmount !== null ? overrideAmount : baseBudget;
        const budgetResult = effectiveBudget !== null ? spent - effectiveBudget : null;

        return {
          id: category.id,
          name: category.name,
          nameHe: category.nameHe || null,
          icon: category.icon,
          color: category.color,
          parentCategoryId: category.parentCategoryId || null,
          baseBudget,
          override: overrideAmount,
          effectiveBudget,
          spent,
          result: budgetResult
        };
      }));

      ctx.body = { success: true, categories: result };
    } catch (error) {
      console.error('getMonthBudget error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to retrieve month budget' };
    }
  }

  /**
   * App: PUT /app/budget/base
   * Body: { expenseCategoryId, householdId, amount }
   * Sets the monthly base budget on the category itself (persists every month).
   */
  async setBaseBudget(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { expenseCategoryId, householdId, amount } = ctx.request.body;

      if (!expenseCategoryId || !householdId || amount === undefined || amount === null) {
        ctx.status = 400;
        ctx.body = { error: 'expenseCategoryId, householdId, and amount are required' };
        return;
      }

      const membership = await HouseholdMember.findOne({
        where: { householdId, appUserId: appUser.id }
      });
      if (!membership) {
        ctx.status = 403;
        ctx.body = { error: 'Not a member of this household' };
        return;
      }

      const category = await ExpenseCategory.findOne({
        where: { id: expenseCategoryId, householdId }
      });
      if (!category) {
        ctx.status = 404;
        ctx.body = { error: 'Category not found' };
        return;
      }

      await category.update({ monthlyBudget: amount >= 0 ? amount : null });

      ctx.body = { success: true };
    } catch (error) {
      console.error('setBaseBudget error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to set base budget' };
    }
  }

  /**
   * App: PUT /app/budget/override
   * Body: { expenseCategoryId, householdId, year, month, amount }
   * Upserts a CategoryBudgetOverride.
   */
  async setOverride(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { expenseCategoryId, householdId, year, month, amount } = ctx.request.body;

      if (!expenseCategoryId || !householdId || !year || !month || amount === undefined || amount === null) {
        ctx.status = 400;
        ctx.body = { error: 'expenseCategoryId, householdId, year, month, and amount are required' };
        return;
      }

      const membership = await HouseholdMember.findOne({
        where: { householdId, appUserId: appUser.id }
      });
      if (!membership) {
        ctx.status = 403;
        ctx.body = { error: 'Not a member of this household' };
        return;
      }

      const [override] = await CategoryBudgetOverride.upsert({
        expenseCategoryId,
        householdId,
        year,
        month,
        amount
      });

      ctx.body = { success: true, override };
    } catch (error) {
      console.error('setOverride error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to set budget override' };
    }
  }

  /**
   * App: GET /app/budget/by-week?householdId=X&year=Y&month=M[&expenseCategoryId=C]
   * Returns spending grouped by ISO week for the calendar month.
   */
  async getSpendingByWeek(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { householdId, year, month, expenseCategoryId } = ctx.query;

      if (!householdId || !year || !month) {
        ctx.status = 400;
        ctx.body = { error: 'householdId, year, and month are required' };
        return;
      }

      const membership = await HouseholdMember.findOne({
        where: { householdId, appUserId: appUser.id }
      });
      if (!membership) {
        ctx.status = 403;
        ctx.body = { error: 'Not a member of this household' };
        return;
      }

      const yearInt = parseInt(year, 10);
      const monthInt = parseInt(month, 10);

      const startDay = await getHouseholdStartDay(householdId);
      const period = getFinancialPeriodForAnchor(yearInt, monthInt, startDay);
      const monthStart = period.start;
      const monthEnd = period.end;

      const where = {
        householdId,
        dateTime: { [Op.between]: [monthStart, monthEnd] }
      };
      if (expenseCategoryId) {
        where.expenseCategoryId = expenseCategoryId;
      }

      const rows = await Expense.findAll({
        attributes: [
          [fn('WEEK', col('dateTime'), 3), 'isoWeek'],
          [fn('SUM', col('amount')), 'total']
        ],
        where,
        group: [fn('WEEK', col('dateTime'), 3)],
        order: [[fn('WEEK', col('dateTime'), 3), 'ASC']],
        raw: true
      });

      // Calculate week labels relative to the first ISO week of the month
      const firstWeek = rows.length > 0 ? parseInt(rows[0].isoWeek, 10) : null;
      const weeks = rows.map((row, idx) => ({
        weekLabel: `W${idx + 1}`,
        total: parseFloat(row.total) || 0
      }));

      ctx.body = { success: true, weeks };
    } catch (error) {
      console.error('getSpendingByWeek error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to retrieve spending by week' };
    }
  }

  /**
   * App: GET /app/budget/by-month?householdId=X&year=Y&startMonth=S&endMonth=E[&expenseCategoryId=C]
   * Returns spending grouped by month for a range.
   */
  async getSpendingByMonth(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { householdId, year, startMonth, endMonth, expenseCategoryId } = ctx.query;

      if (!householdId || !year || !startMonth || !endMonth) {
        ctx.status = 400;
        ctx.body = { error: 'householdId, year, startMonth, and endMonth are required' };
        return;
      }

      const membership = await HouseholdMember.findOne({
        where: { householdId, appUserId: appUser.id }
      });
      if (!membership) {
        ctx.status = 403;
        ctx.body = { error: 'Not a member of this household' };
        return;
      }

      const yearInt = parseInt(year, 10);
      const startMonthInt = parseInt(startMonth, 10);
      const endMonthInt = parseInt(endMonth, 10);

      const startDay = await getHouseholdStartDay(householdId);

      // Build the list of financial period anchors covering startMonth..endMonth.
      // Anchors may straddle a year boundary when startMonth > endMonth.
      const anchors = [];
      let curYear = yearInt;
      let curMonth = startMonthInt;
      while (true) {
        anchors.push({ year: curYear, month: curMonth });
        if (curYear === yearInt && curMonth === endMonthInt) break;
        curMonth += 1;
        if (curMonth > 12) { curMonth = 1; curYear += 1; }
        if (anchors.length > 36) break; // safety
      }

      const monthNames = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
        'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

      const months = await Promise.all(anchors.map(async ({ year: yr, month: mo }) => {
        const period = getFinancialPeriodForAnchor(yr, mo, startDay);
        const where = {
          householdId,
          dateTime: { [Op.between]: [period.start, period.end] }
        };
        if (expenseCategoryId) {
          where.expenseCategoryId = expenseCategoryId;
        }
        const row = await Expense.findOne({
          attributes: [[fn('SUM', col('amount')), 'total']],
          where,
          raw: true
        });
        return {
          label: `${monthNames[mo - 1]} ${yr}`,
          total: parseFloat(row?.total) || 0
        };
      }));

      ctx.body = { success: true, months };
    } catch (error) {
      console.error('getSpendingByMonth error:', error);
      ctx.status = 500;
      ctx.body = { error: 'Failed to retrieve spending by month' };
    }
  }
  // ── Budget Plan Items ───────────────────────────────────────────────────────

  /**
   * GET /app/budget/plan-items?householdId=X&year=Y&month=M
   */
  async getPlanItems(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { householdId, year, month } = ctx.query;
      if (!householdId || !year || !month) {
        ctx.status = 400; ctx.body = { error: 'householdId, year, month required' }; return;
      }
      const membership = await HouseholdMember.findOne({ where: { householdId, appUserId: appUser.id } });
      if (!membership) { ctx.status = 403; ctx.body = { error: 'Not a member' }; return; }

      const items = await BudgetPlanItem.findAll({
        where: { householdId, year: parseInt(year), month: parseInt(month) },
        order: [['createdAt', 'ASC']]
      });
      ctx.body = { success: true, items };
    } catch (err) {
      console.error('getPlanItems error:', err);
      ctx.status = 500; ctx.body = { error: 'Failed to get plan items' };
    }
  }

  /**
   * POST /app/budget/plan-items
   * Body: { householdId, expenseCategoryId, year, month, description, amount }
   */
  async createPlanItem(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { householdId, expenseCategoryId, year, month, description, minAmount, maxAmount } = ctx.request.body;
      if (!householdId || !expenseCategoryId || !year || !month) {
        ctx.status = 400; ctx.body = { error: 'householdId, expenseCategoryId, year, month required' }; return;
      }
      const membership = await HouseholdMember.findOne({ where: { householdId, appUserId: appUser.id } });
      if (!membership) { ctx.status = 403; ctx.body = { error: 'Not a member' }; return; }

      const item = await BudgetPlanItem.create({
        householdId, expenseCategoryId, year, month,
        description: description || null,
        minAmount: parseFloat(minAmount) || 0,
        maxAmount: parseFloat(maxAmount) || 0
      });
      ctx.status = 201;
      ctx.body = { success: true, item };
    } catch (err) {
      console.error('createPlanItem error:', err);
      ctx.status = 500; ctx.body = { error: 'Failed to create plan item' };
    }
  }

  /**
   * PUT /app/budget/plan-items/:id
   * Body: { description, amount }
   */
  async updatePlanItem(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { id } = ctx.params;
      const { description, minAmount, maxAmount } = ctx.request.body;

      const item = await BudgetPlanItem.findByPk(id);
      if (!item) { ctx.status = 404; ctx.body = { error: 'Item not found' }; return; }

      const membership = await HouseholdMember.findOne({ where: { householdId: item.householdId, appUserId: appUser.id } });
      if (!membership) { ctx.status = 403; ctx.body = { error: 'Not a member' }; return; }

      await item.update({
        description: description !== undefined ? (description || null) : item.description,
        minAmount: minAmount !== undefined ? (parseFloat(minAmount) || 0) : item.minAmount,
        maxAmount: maxAmount !== undefined ? (parseFloat(maxAmount) || 0) : item.maxAmount
      });
      ctx.body = { success: true, item };
    } catch (err) {
      console.error('updatePlanItem error:', err);
      ctx.status = 500; ctx.body = { error: 'Failed to update plan item' };
    }
  }

  /**
   * DELETE /app/budget/plan-items/:id
   */
  async deletePlanItem(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { id } = ctx.params;

      const item = await BudgetPlanItem.findByPk(id);
      if (!item) { ctx.status = 404; ctx.body = { error: 'Item not found' }; return; }

      const membership = await HouseholdMember.findOne({ where: { householdId: item.householdId, appUserId: appUser.id } });
      if (!membership) { ctx.status = 403; ctx.body = { error: 'Not a member' }; return; }

      await item.destroy();
      ctx.body = { success: true };
    } catch (err) {
      console.error('deletePlanItem error:', err);
      ctx.status = 500; ctx.body = { error: 'Failed to delete plan item' };
    }
  }

  // ── Budget Month Config ─────────────────────────────────────────────────────

  /**
   * GET /app/budget/month-config?householdId=X&year=Y&month=M
   */
  async getMonthConfig(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { householdId, year, month } = ctx.query;
      if (!householdId || !year || !month) {
        ctx.status = 400; ctx.body = { error: 'householdId, year, month required' }; return;
      }
      const membership = await HouseholdMember.findOne({ where: { householdId, appUserId: appUser.id } });
      if (!membership) { ctx.status = 403; ctx.body = { error: 'Not a member' }; return; }

      const config = await BudgetMonthConfig.findOne({
        where: { householdId, year: parseInt(year), month: parseInt(month) }
      });
      ctx.body = { success: true, config: config || null };
    } catch (err) {
      console.error('getMonthConfig error:', err);
      ctx.status = 500; ctx.body = { error: 'Failed to get month config' };
    }
  }

  /**
   * PUT /app/budget/month-config
   * Body: { householdId, year, month, startAmount }
   */
  async upsertMonthConfig(ctx) {
    try {
      const appUser = ctx.state.appUser;
      const { householdId, year, month, startAmount, expectedIncome } = ctx.request.body;
      if (!householdId || !year || !month) {
        ctx.status = 400; ctx.body = { error: 'householdId, year, month required' }; return;
      }
      const membership = await HouseholdMember.findOne({ where: { householdId, appUserId: appUser.id } });
      if (!membership) { ctx.status = 403; ctx.body = { error: 'Not a member' }; return; }

      const [config] = await BudgetMonthConfig.upsert({
        householdId, year, month,
        startAmount: startAmount !== undefined ? (parseFloat(startAmount) || null) : null,
        expectedIncome: expectedIncome !== undefined ? (parseFloat(expectedIncome) || null) : null
      });
      ctx.body = { success: true, config };
    } catch (err) {
      console.error('upsertMonthConfig error:', err);
      ctx.status = 500; ctx.body = { error: 'Failed to save month config' };
    }
  }
}

module.exports = new BudgetsController();
